rust/src/write.rs

```diff
@@ -1545,26 +1585,45 @@ impl<W: Write + Seek> ChunkWriter<W> {
             hasher.combine(compressed_crc);
         }
         let mut writer = CountingCrcWriter::with_hasher(writer, post_chunk_header_crc);
-
         // Write our message indexes
-        let data_end = writer.stream_position()?;
+        let data_end = match writer.stream_position() {
+            Ok(v) => v,
+            Err(err) => {
+                return Err((writer.finalize().0, err.into()));
+            }
+        };
```

Comment:
nit: if you wanted you could use a macro for these:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=e131ddb259c38f3fb68e16318f559a32